### PR TITLE
Fix racy CopyOnWrite data structure serialization write

### LINKED ISSUE #18129: Serialization of CopyOnWriteArrayList is racy
The `CopyOnWriteArrayListStreamSerializer` delegates the serialization to the superclass, which is racy. It first writes the size and then the elements, but the collection can be modified in the meantime. Also the deserialization adds the elements one by one, which causes the copy-on-write.

Causes hazelcast/hazelcast-jet#2864.

### PR DESCRIPTION
**Pull request by ramizdundar**
As discussed in #18129 all concurrent data structures are prone to race conditions when serialized if they are not cloned. However due to nature of the COW data structures there are ways to make them serialize without cloning. For example pseudocode in #18129. In order to not break backward compatibility, this PR uses splititerators to get size directly from the iterator.

Fixes #18129

### CODE REVIEW

**Review by sancar — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/defaultserializers/CopyOnWriteArrayListStreamSerializer.java [hunk 1]
```diff
@@ -42,4 +45,18 @@ public int getTypeId() {
 
         return new CopyOnWriteArrayList<>(collection);
     }
+
+    @Override
+    @SuppressWarnings("DuplicatedCode")
+    public void write(ObjectDataOutput out, CopyOnWriteArrayList<E> collection) throws IOException {
+        Spliterator<E> cowSplitIterator = collection.spliterator();
+        out.writeInt((int) cowSplitIterator.getExactSizeIfKnown());
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/defaultserializers/CopyOnWriteArrayListStreamSerializer.java:53]: `getExactSizeIfKnown` can return -1 according to the contract. I guess it won't for COWAL, but we can assert that.

**Review by viliam-durina — APPROVED** [review]
Approving with a minor comment.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/defaultserializers/CopyOnWriteArrayListStreamSerializer.java (hunk 1 above)
Comment (ramizdundar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/defaultserializers/CopyOnWriteArrayListStreamSerializer.java]: Done https://github.com/hazelcast/hazelcast/pull/18285/commits/4e1a00c4b0aa682df12c7bc2ec8d252df9acbb00

Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/defaultserializers/CopyOnWriteArrayListStreamSerializer.java]: Contract of CopyOnWriteArrayList states that this never return -1. 
So this behaviour can not ever change.

**Pull request merged**